README.md
# [Baremetrics](https://baremetrics.com/) Select Search Filter
_[Baremetrics](https://baremetrics.com) provides one-click analytics & insights for Stripe. **[Get started today!](https://baremetrics.com)**_

---

The Baremetrics custom selection picker is a glorified select box solution. It's super flexible and allows many different types and ranges of complex selections while at the end of the day generating a very simple, concise JSON object which you can use for whatever.

Design by [Chris Meeks](https://dribbble.com/ChrisMeeks)  
Code by [Tyler van der Hoeven](https://github.com/tyvdh)

[View a demo](http://baremetrics.github.io/select)  
<!--[View in a live production app](https://demo.baremetrics.com/sales/search)-->

![screenshot2](http://tyler.link/dxB4/Screen%20Shot%202015-11-30%20at%201.14.52%20PM.png)
![screenshot1](http://tyler.link/dOVh/Screen%20Shot%202015-09-30%20at%203.51.46%20PM.png)
![screenshot3](http://tyler.link/dOe2/Screen%20Shot%202015-09-30%20at%203.51.00%20PM.png)

## Installing
Using the select filter tags is super straight forward, you just need to target a div with the class of `bm-select` where your filter options will generate.

We've also got the option to spit out a separate filter cloud for non selectable results. That too needs a `bm-select` class.

In the demo I'm using:
```html
<div class="select bm-select"></div>
<!--
<button class="reset">Reset</button>
<button class="submit">Submit</button>
-->
<div class="result bm-select"></div>
```

Once you've got the div's ready all you've gotta do is create an instance of the `Select` object with all the tags you'd like to generate.

```js
var filters = new Select({
  cloud: '.select',
  tags: [
    {
      title: 'Name',
      type: 'bmInput',
      options: [
        {
          title: 'Contains',
          query: 'with',
          selected: 'Metrics',
        },
        {
          title: 'Doesn\'t Contain',
          query: 'without',
        }
      ]
    },
    {
      title: 'Price',
      type: 'bmInput',
      // selected: '$129.99',
      placeholder: 'e.g. $99.99'
    },
    {
      title: 'Tech',
      type: 'bmMulti',
      options: [
        {
          title: 'Python',
          selected: true,
        },
        {
          title: 'MongoDB',
          selected: true,
        },
        {
          title: 'Node',
        },
        {
          title: 'PHP',
        },
        {
          title: 'Meteor',
        },
        {
          title: 'MySQL',
        },
        {
          title: 'Ruby',
        }
      ]
    },
    {
      title: 'Billing Interval',
      options: [
        {
          title: 'Annual'
        },
        {
          title: 'Monthly'
        }
      ]
    }
  ]
});

// Also just for kicks here are a couple actions
// I'm using in the demo to reset the filter selections
// and spit out the result html

$('.reset').click(function() {

  // Resets all the filters to 'None'
  filters.reset();
});

$('.submit').click(function() {
  
  // This gives you the selected tag JSON
  var filters_array = filters.call();

  // Clears out any existing tags in the result cloud
  var result_cloud = $('.result').empty();

  // Loops over the results and adds each one to the result cloud
  $.each(filters_array, function(i, d) {
    var HTML = filters.addResultHTML(d);
    result_cloud.append(HTML);
  });
});
```

## Params

### Layer One (`cloud` & `tag`)
`Select` allows for one object argument with two keys, `cloud` and `tags`.

- `cloud:`
  - The identifier for where we should generate the select filter tags
- `tags:`
  - An array of tag objects

### Layer Two (`title`, `type` & `options`)
A tag object has three keys. `title`, `type` and `options`

- `title:`
  - The title of the tag
- `type:`
  - One of either `'bmInput'`, `'bmMulti'` or `'bmSingle'` 
  - Default is `'bmSingle'`
- `options:`
  - Array of dropdown objects for this tag
- `placeholder:`
  - If `bmInput` without an options array you can set a placeholder for the input field
- `selected:`
  - If `bmInput` without an options array  you can set a pre-selected value for the input field

### Layer Three (`title`, `query` & `selected`)
The options array has three allowable keys. `title`, `query` or `selected`

- `title:` 
  - The title for the dropdown item
- `query:`
  - The shorthand name for this selection. 
  - Defaults to lowercase alpha char regex
- `selected:`
  - If `'bmInput'`, this is the value to pre-fill the input div with
  - Otherwise it's just a boolean to pre-select this item or not 
  - Default obviously is `false`


## Functions
Now that we're all set up with filters, how do we access the values for use throughout the rest of our application? Easy!

### `[Select instance].reset()`
First and easiest is `[Select instance].reset()` which will just clear all filters of their current selections. Note however this does not automatically reset the values unless a `[Select instance].call()` is also called. This function just visually clears the playing field.

### `[Select instance].call()`
The second function is `[Select instance].call()`. This will return a clone of the original tags array you passed to the `Select` instance. The only thing which will be updated is where the `selected` keys and values are. Thus you could save this return array back to subsequent calls of the `Select` object as a sort of "what did I search for last" type deal. We do this exact thing in the Baremetrics app using `localStorage`. The more common use though obviously is to use this as your option in a call back to the server for use in some type of API call.

### `[Select instance].addResultHTML([one_tag_object])`
The last function you'll likely want to make use of is the `[Select instance].addResultHTML([one_tag_object])`. What this allows you to do is pass in one of the tag objects from either the `call()` return or even the original object and generate a sort of static or dumb tag. In the [demo](http://baremetrics.github.io/select/) this is the function that is called when you click the "Submit" button. It only takes one object at a time though so you'll likely want to create an each loop if you plan on generating a tag cloud of all the return search objects.

---

## Developing

I've included my signature gulpfile too so be sure and take a look at that as well.

```bash
$ cd <project directory>
$ npm install
$ gulp
```

I also use [pow](http://pow.cx/) and the [powder gem](https://github.com/Rodreegez/powder) to run my local dev environments but however you plan on wrangling that the gulpfile turns on a livereload server so as long as you have the files serving somehow any changes you make will show up instantly.

## Dependencies
- [jQuery](https://jquery.com/)


changelog.md
## v1.0.1 – "Single"

Added two new keys for the `bmInput` tag. `placeholder` and `selected`.
Basically it's a shortcut for allowing you to set a tag without any optional modifiers.

```js
...
{
  title: 'Price',
  type: 'bmInput',
  selected: '$129.99',
  placeholder: 'e.g. $99.99'
}
...
```

![screen shot 2015-11-30 at 1 10 54 pm](https://cloud.githubusercontent.com/assets/4383610/11480143/d13b2ec4-9763-11e5-9f27-2be1cdd5bd45.png)

js/Select.js
'use strict';

function Select(options) {
  if (!options) 
    return true;

  this.cloud = $(options.cloud);
  this.tags = options.tags;

  var self = this;

  // Make the tags
  $(this.tags).each(function(i, d) {
    self.addSelectHTML(d);
  });
  
  $('.bm-tag', this.cloud).each(function(i) {
    self.addSelectActions($(this)); // Add actions to the new tags
    self.addSelectValues($(this), self.tags[i].options || [{selected: self.tags[i].selected}]); // Add values to the new tags
  });
}

Select.prototype.addSelectActions = function(tag) {
  var self = this;

  $('.bm-title, .bm-drop', tag).off('click').click(function() {
    var target = $(this).hasClass('bm-drop') ? $(this) : $('.bm-drop', tag).first();
    var dropdown_type = $('.bm-dropdown', target).data('type');
    var input = $('.bm-input', tag);

    if (!target.hasClass('bm-open')) {
      self.closeDrop(tag);
      target.addClass('bm-open');
      $('.bm-dropdown', target).show();
    } else {
      self.closeDrop(tag);
    }

    if ($('.bm-item', target).length) {
      $('.bm-item', target).off('click').click(function() {
        var new_value = $(this).html();
        var new_query = $(this).data('query');

        $('span', target).data('query', new_query).html(new_value);
        target.removeClass('bm-more');
        tag.addClass('bm-active');

        if (dropdown_type == 'input') {
          target.removeClass('bm-inner');

          if (new_query == 'all') {
            input.hide();
          } else {
            target.addClass('bm-inner');
            input
              .show()
              .off('focus')
              .focus()
              .off('focus')
              .focus(function() {self.closeDrop(tag)});
          }
        }

        else if (dropdown_type == 'multi') {
          if (new_query != 'all' && !target.next().length) {
            var used_tags = $('.bm-drop', tag).length;
            var unused_tags = $('.bm-dropdown:first li:nth-child(1n+2)', tag).length;
            var drop_clone = $(target).clone();

            if (used_tags != unused_tags) {
              drop_clone.removeClass('bm-open').addClass('bm-more');
              drop_clone.find('.bm-dropdown').hide();
              drop_clone.find('span').html('More');

              target.addClass('bm-inner');
              $(target).after(drop_clone);
              self.addSelectActions(tag);
            } else {
              $('span', target).data('query', 'all').html('All');
              $('.bm-drop', tag).not(target).remove();
              $('.bm-dropdown li', tag).removeClass('bm-disabled');

              return true;
            }

            $('.bm-drop span', tag).each(function() {
              var query = $(this).data('query');
              $('.bm-dropdown', tag).find('[data-query="'+ query +'"]').addClass('bm-disabled');
            });
          } else if (new_query == 'all') {
            target.removeClass('bm-inner');
            $('.bm-drop', tag).not(target).remove();
            $('.bm-dropdown li', tag).removeClass('bm-disabled');
          } else {
            $('.bm-dropdown li', tag).removeClass('bm-disabled');

            $('.bm-drop span', tag).each(function() {
              var query = $(this).data('query');
              $('.bm-dropdown', tag).find('[data-query="'+ query +'"]').addClass('bm-disabled');
            });
          }
        }
      });
    } else {
      input.focus();
    }

    self.cloud.off('click').click(function(event) {
      $('html').one('click', function() {
        self.closeDrop(tag);
      });

      event.stopPropagation();
    });
  });

  $('.bm-close', tag).off('click').click(function() {
    var target = $(this).parent();
    var prev = target.prev();
    var next = target.next();
    var query = target.find('span').data('query');
    var dropdown_type = $('.bm-dropdown', target).data('type');

    target.parent().find('.bm-dropdown').find('[data-query="'+ query +'"]').removeClass('bm-disabled');

    if (dropdown_type == 'multi') {
      if (prev.hasClass('bm-title') && (next.hasClass('bm-more') || !next.length)) {
        next.remove();
        target.parent().removeClass('bm-active');
        target.removeClass('bm-inner').find('span').data('query', 'none').html('None');
      } else if (!target.parent().find('.bm-more').length) {
        var target_clone = target.clone();

        target_clone.removeClass('bm-inner').addClass('bm-more');
        target_clone.find('span').html('More');

        target.parent().find('.bm-drop:last').after(target_clone);
        target_clone.prev().addClass('bm-inner');
        target.remove();

        self.addSelectActions(tag);
      } else {
        target.remove();
      }
    } else if (dropdown_type == 'input') {
      next.hide();
      target.parent().removeClass('bm-active');
      target.removeClass('bm-inner').find('span').data('query', 'none').html('None');
    } else {
      next.remove();
      target.parent().removeClass('bm-active');
      target.removeClass('bm-inner').find('span').data('query', 'none').html('None');
    }
  });

  $('[contenteditable]', tag).off('keyup keydown input').on('keyup keydown input', function() {
    var target = $(this).parent();
    var str = $(this).html().replace(/<br\s*[\/]?>/gi, '');

    if (str.length > 0) {
      target.has('.bm-drop').length ? false : target.addClass('bm-active');
      $(this).removeClass('bm-empty');
    } else {
      target.has('.bm-drop').length ? false : target.removeClass('bm-active');
      $(this).empty().addClass('bm-empty');
    }
  });
}

Select.prototype.addSelectValues = function(tag, options) {
  if (options) {
    $.each(options, function(i, d) {
      if (d.selected) {
        if (d.query || d.title) {
          var trigger = $('.bm-drop', tag).last().trigger('click');
          var query = d.query || d.title.replace(/[^a-zA-Z]/g, '').toLowerCase();

          $('[data-query="'+ query +'"]', trigger).trigger('click');
        } else {
          tag.addClass('bm-active');
        }

        $('.bm-input', tag).removeClass('bm-empty').html(d.selected);
      }
    });
  }

  window.setTimeout(function() {
    $('.bm-input').blur();
  }, 0);
}

Select.prototype.closeDrop = function(tag) {
  $('.bm-drop', tag).removeClass('bm-open');
  $('.bm-dropdown', tag).hide();
}

Select.prototype.addSelectHTML = function(object) {
  var HTML = '<div class="bm-tag">'+
    '<div class="bm-title bm-cell">'+ object.title +'</div>';

  var type = object.type ? this[object.type]() : this.bmSingle();

  if (object.options) {
    HTML += '<div class="bm-drop bm-cell"><i class="bm-close"></i> <span data-query="none">None</span> <i class="bm-caret"></i><ul class="bm-dropdown" data-type="'+ type +'" style="display:none;">';

    $.each(object.options, function(i, d) {
      if (i == 0 && type == 'multi') {
        HTML += '<li class="bm-item" data-query="all">All</li>';
      }
      
      var query = d.query || d.title.replace(/[^a-zA-Z]/g, '').toLowerCase();

      HTML += '<li class="bm-item" data-query="'+ query +'">'+ d.title +'</li>';
    });

    HTML += '</ul></div>';
  }

  if (type == 'input') {
    HTML += '<div class="bm-input bm-empty bm-cell" placeholder="'+ (object.placeholder || 'Enter value...') +'"'+ (object.options ? 'style="display:none;"' : '') +'contenteditable></div>';
  }

  HTML += '</div>';

  $(this.cloud).append(HTML);
}

Select.prototype.addResultHTML = function(object) {
  var options = $.map(object.options || [{selected: object.selected}], function(d) {
    return {
      title: d.title,
      selected: d.selected || false
    }
  }).filter(function(e) {return e.selected != false});

  if (!options.length)
    return false;

  var HTML = '<div class="bm-tag bm-final">'+
    '<div class="bm-title bm-cell">'+ object.title +'</div>';

  if (this[object.type] == this.bmInput) {
    if (options[0].title)
      HTML += '<div class="bm-modifier bm-cell">'+ options[0].title +'</div>';

    HTML += '<div class="bm-values bm-cell">'+ (Number(options[0].selected) ? options[0].selected : '"'+ options[0].selected +'"') +'</div>';
  } else {
    HTML += '<div class="bm-values bm-cell">';

    $.each(options, function(i, d) {
      HTML += (i == 0 ? '' : ', ') + d.title;
    });

    HTML += '</div>';
  }

  HTML += '</div>';

  return HTML;
}

Select.prototype.call = function() {
  var self = this;
  var tags = self.tags;

  $('.bm-tag', this.cloud).each(function(i, d) {
    var title = $('.bm-title', d).html();
    var index = tags.map(function(e){return e.title}).indexOf(title);
    var selected = $.makeArray($('.bm-drop:not(.bm-more) span', d).map(function() {
      return $(this).data('query');
    }));

    if (tags[index].options) {
      $.each(tags[index].options, function(j, e) {
        var query = e.query || e.title.replace(/[^a-zA-Z]/g, '').toLowerCase();

        if (selected.indexOf(query) != -1 || selected[0] == 'all') {
          var input = $('.bm-input', d).html();

          tags[index].options[j].selected = input || true;
        } else {
          tags[index].options[j].selected = false;
        }
      });
    } else {
      var input = $('.bm-input', d).html();
      tags[index].selected = input || false
    }
  });

  return tags;
}

Select.prototype.reset = function() {
  $('.bm-active', this.cloud).each(function() {
    var d = $(this).removeClass('bm-active');
    
    $('.bm-drop', d).not(':first').remove();
    $('.bm-dropdown li', d).removeClass('bm-disabled');

    d.has('.bm-drop').length ? 
      $('.bm-input', d).hide() : 
      $('.bm-input', d).addClass('bm-empty').empty();

    $('.bm-drop:first', d).removeClass('bm-active bm-inner').find('span').data('query', 'none').html('None');
  });
}

Select.prototype.bmInput = function() {
  return 'input';
}

Select.prototype.bmMulti = function() {
  return 'multi';
}

Select.prototype.bmSingle = function() {
  return 'single';
}

js/app.js
'use strict';

$(function() {
  var filters = new Select({
    cloud: '.select',
    tags: [
      {
        title: 'Name',
        type: 'bmInput',
        options: [
          {
            title: 'Contains',
            query: 'with',
            selected: 'Metrics',
          },
          {
            title: 'Doesn\'t Contain',
            query: 'without',
          }
        ]
      },
      {
        title: 'Price',
        type: 'bmInput',
        // selected: '$129.99',
        placeholder: 'e.g. $99.99'
      },
      {
        title: 'Tech',
        type: 'bmMulti',
        options: [
          {
            title: 'Python',
            selected: true,
          },
          {
            title: 'MongoDB',
            selected: true,
          },
          {
            title: 'Node',
          },
          {
            title: 'PHP',
          },
          {
            title: 'Meteor',
          },
          {
            title: 'MySQL',
          },
          {
            title: 'Ruby',
          }
        ]
      },
      {
        title: 'Billing Interval',
        options: [
          {
            title: 'Annual'
          },
          {
            title: 'Monthly'
          }
        ]
      }
    ]
  });
  
  $('.reset').click(function() {
    filters.reset();
  });

  $('.submit').click(function() {
    var filters_array = filters.call();
    var result_cloud = $('.result').empty();

    $.each(filters_array, function(i, d) {
      var HTML = filters.addResultHTML(d);
      result_cloud.append(HTML);
    });
  });
});